### PR TITLE
Disable ipv4 and ipv6 protocol.

### LINKED ISSUE #33647: nmcli: Add support for setting IP address to disabled/ignore for IPv4 and IPv6 retrospectively.
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
nmcli

##### ANSIBLE VERSION
```
ansible 2.4.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/momcilo/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.14 (default, Nov  2 2017, 19:02:24) [GCC 7.2.1 20170829 (Red Hat 7.2.1-1)]
```

##### CONFIGURATION
No changes.

##### OS / ENVIRONMENT
Fedora/Centos

##### SUMMARY
Add support for setting IP address to disabled/ignore for IPv4 and IPv6 retrospectively.

##### STEPS TO REPRODUCE
Use special keyword 'disabled' to set new ip4 or ip6 to none.

```yaml
- hosts: localhost

  tasks:

  - name: Configure ethernet bridge to OVS
    nmcli:
     state: present
     conn_name: enp1s0f0
     ip4: disabled
```

##### EXPECTED RESULTS
Connection gets altered so any IP addresses are removed.

##### ACTUAL RESULTS
Not yet implemented.

### PR DESCRIPTION
**Pull request by amolkahat**
##### SUMMARY
It will disable the IPv4 or IPv6 protocol if the user does not want it.
Added code for disable ipv4 and ipv6 protocol.
Fixes: #33647

Signed-off-by: Amol Kahat <akahat@redhat.com>

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
nmcli

##### ANSIBLE VERSION

```
2.5devel
```

### CODE REVIEW

File: lib/ansible/modules/net_tools/nmcli.py [hunk 1]
```diff
@@ -1182,6 +1182,17 @@ def modify_connection(self):
             cmd = self.modify_connection_bridge_slave()
         elif self.type == 'vlan':
             cmd = self.modify_connection_vlan()
+            # If no dns given, then we assume that IPv4/IPv6 is not used on the interface
+            # so we are going to disable the protocol, otherwise interface won't come up.
+
+            if self.ip4 is 'disabled':
+                self.execute_command(cmd)
+                cmd = self.modify_conn_disable_ipv4()
```
Comment (Akasurde) [inline lib/ansible/modules/net_tools/nmcli.py]: You are not executing command returned by `modify_conn_disable_ipv4` so I am sure that it has no effect.

File: lib/ansible/modules/net_tools/nmcli.py [hunk 2]
```diff
@@ -1182,6 +1182,17 @@ def modify_connection(self):
...
+
+            if self.ip6 is 'ignore':
+                self.execute_command(cmd)
```
Comment (Akasurde) [inline lib/ansible/modules/net_tools/nmcli.py:1452]: idem

File: lib/ansible/modules/net_tools/nmcli.py [hunk 3]
```diff
@@ -1182,6 +1182,17 @@ def modify_connection(self):
...
         elif self.type == 'vlan':
             cmd = self.modify_connection_vlan()
+            # If no dns given, then we assume that IPv4/IPv6 is not used on the interface
+            # so we are going to disable the protocol, otherwise interface won't come up.
+
+            if self.ip4 is 'disabled':
+                self.execute_command(cmd)
+                cmd = self.modify_conn_disable_ipv4()
+
+            if self.ip6 is 'ignore':
```
Comment (Akasurde) [inline lib/ansible/modules/net_tools/nmcli.py]: Does this functionality only affects VLAN ? What about other connection types and modes like ethernet and bridge etc.,?

**Review by Akasurde — CHANGES_REQUESTED** [review]
Integration tests are missing.

**Comment by Akasurde** [comment]
@Colins-Git Could you please take a look ?

File: lib/ansible/modules/net_tools/nmcli.py (hunk 1 above)
Comment (Akasurde) [reply lib/ansible/modules/net_tools/nmcli.py]: My bad. Didn't see if in later stages.

File: lib/ansible/modules/net_tools/nmcli.py (hunk 2 above)
Comment (Akasurde) [reply lib/ansible/modules/net_tools/nmcli.py:1452]: My bad. Didn't see if in later stages.

File: lib/ansible/modules/net_tools/nmcli.py [hunk 4]
```diff
@@ -1182,6 +1182,17 @@ def modify_connection(self):
             cmd = self.modify_connection_bridge_slave()
         elif self.type == 'vlan':
             cmd = self.modify_connection_vlan()
+            # If no dns given, then we assume that IPv4/IPv6 is not used on the interface
+            # so we are going to disable the protocol, otherwise interface won't come up.
+
+            if self.ip4 is 'disabled':
```
Comment (thomwiggers) [inline lib/ansible/modules/net_tools/nmcli.py]: You shouldn't do string comparison with `is`.

https://stackoverflow.com/questions/1504717/why-does-comparing-strings-in-python-using-either-or-is-sometimes-produce

File: lib/ansible/modules/net_tools/nmcli.py (hunk 3 above)
Comment (thomwiggers) [reply lib/ansible/modules/net_tools/nmcli.py]: Again string comparison with `is`.

Comment (Akasurde) [reply lib/ansible/modules/net_tools/nmcli.py]: Good catch @thomwiggers

**Comment by amolkahat** [comment]
@maxamillion Could you please review it?

**Comment by maxamillion** [comment]
@amolkahat I'm out on Paternity Leave for the next 6 weeks and won't be online much, if you need this done quickly I'd suggest pinging someone else.

File: lib/ansible/modules/net_tools/nmcli.py [hunk 5]
```diff
@@ -1199,12 +1199,31 @@ def modify_connection(self):
             cmd = self.modify_connection_vlan()
         elif self.type == 'generic':
             cmd = self.modify_connection_ethernet(conn_type='generic')
+
         if cmd:
+            # If no dns given, then we assume that IPv4/IPv6 is not used on the interface
+            # so we are going to disable the protocol, otherwise interface won't come up.
+
+            if self.ip4 == 'disabled':
```
Comment (Akasurde) [inline lib/ansible/modules/net_tools/nmcli.py:1447]: Can you make option common for ipv4 and ipv6 same ?

Comment (Akasurde) [inline lib/ansible/modules/net_tools/nmcli.py:1447]: Add an example for this in EXAMPLE section.

File: lib/ansible/modules/net_tools/nmcli.py [hunk 6]
```diff
@@ -1199,12 +1199,31 @@ def modify_connection(self):
...
+
+            if self.ip6 == 'ignore':
+                self.execute_command(cmd)
+                cmd = self.modify_conn_disable_ipv6()
             return self.execute_command(cmd)
         else:
             self.module.fail_json(msg="Type of device or network connection is required "
                                       "while performing 'modify' operation. Please specify 'type' as an argument.")
 
+    def modify_conn_disable_ipv4(self):
```
Comment (Akasurde) [inline lib/ansible/modules/net_tools/nmcli.py]: Combine these two methods together with argument.

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Comment by Akasurde** [comment]
@amolkahat Are you still working on this PR ? else, let me know I will take over this PR.

File: lib/ansible/modules/net_tools/nmcli.py [hunk 7]
```diff
@@ -434,6 +434,23 @@
       type: ethernet
       state: present
 
+# To disable the ipv4 connection:
```
Comment (Akasurde) [inline lib/ansible/modules/net_tools/nmcli.py]:
```yaml
- name: To disable the ipv4 connection
  nmcli:
    type: vlan
    conn_name: my-eth1
    ifname: eth1
    ip4: disabled
    state: present

- name: To disable the ipv6 connection
  nmcli:
    type: vlan
    conn_name: my-eth1
    ifname: eth1
    ip6: ignore
    state: present
```

**Comment by Akasurde** [comment]
@ssahani Could you please take over this PR and let me know ? Thanks in advance.

**Comment by ssahani** [comment]
@Akasurde  rather than chosing enable/disable we can support all the settings for ipv4 and ipv6

```
[sus@Zeus ansible]$ nmcli connection modify ip-tunnel-eth0 ipv6.method  [ignore, auto, dhcp, link-local, manual, shared]
```

for example
```
- name:
  nmcli:
    conn_name: my-eth1
    ifname: eth1
    ip6: ignore/auto/dhcp/link-local/manual
    state: present
```

same can be done with ipv4
we parse the value from user and apply . In this way we can apply the whole set what it supports.

**Comment by amolkahat** [comment]
> @Akasurde rather than chosing enable/disable we can support all the settings for ipv4 and ipv6
> 
> ```
> [sus@Zeus ansible]$ nmcli connection modify ip-tunnel-eth0 ipv6.method  [ignore, auto, dhcp, link-local, manual, shared]
> ```
> 
> for example
> 
> ```
> - name:
>   nmcli:
>     conn_name: my-eth1
>     ifname: eth1
>     ip6: ignore/auto/dhcp/link-local/manual
>     state: present
> ```
> 
> same can be done with ipv4
> we parse the value from user and apply . In this way we can apply the whole set what it supports.

Thanks @ssahani. I think its good idea to add other functionalities as well. If its ok, I could add those additional functionalities in subsequent PR's. 

@Akasurde: PTAL

**Comment by arrfab** [comment]
Just wondering the status of this PR, as I wanted also to create bridge over vlan interfaces, dedicated for VMs, (so hypervisor doesn't need ip connectivity at all) and even tried to backport this PR into ansible 2.6 but while it seems to work on modify, at creation time it doesn't.
For the time being I'm just distributing the ifcfg-* files (as jinja2 templates) and just call "nmcli con reload" after, when it was changed.

**Comment by Akasurde** [comment]
Closing due to inactivity. Please feel free to open a new PR. Thanks for the contribution.

**Pull request closed without merging**